src/odbc/odbc.c

```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
-		desc_alloc_records(apd, orig_apd_size);
-		odbc_errs_add(&stmt->errs, "HY001", NULL);
-		ODBC_EXIT_(stmt);
-	}
-	drec = &ipd->records[ipar - 1];
+		if (fSqlType == SQL_SS_TABLE) {
+			/* For a table value, the pcbValue is a buffer containing the number of rows present */
+			/* Replace it with a buffer to the acutal size of a TDS_TVP struct */
+			/* TODO: Replace with a factory method or global variable? */
+			drec->sql_desc_octet_length_ptr = malloc(sizeof(SQLLEN));
+			*(drec->sql_desc_octet_length_ptr) = sizeof(SQLTVP);
 
-	drec->sql_desc_parameter_type = fParamType;
-	if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS) {
-		desc_alloc_records(ipd, orig_ipd_size);
-		desc_alloc_records(apd, orig_apd_size);
-		odbc_errs_add(&stmt->errs, "HY004", NULL);
-		ODBC_EXIT_(stmt);
-	}
-	if (is_numeric) {
-		drec->sql_desc_precision = cbColDef;
-		drec->sql_desc_scale = ibScale;
+			tvp_name_len = cbValueMax == SQL_NTS ? strlen(rgbValue) : cbValueMax;
+			if ((tvp_name = malloc(tvp_name_len + 1)) == NULL) {
+				odbc_errs_add(&stmt->errs, "HY001", NULL);
+				ODBC_EXIT_(stmt);
+			}
+
+			strncpy(tvp_name, rgbValue, tvp_name_len);
+			tvp_name[tvp_name_len] = 0;
+
+			/* Tokenize and extract the schema & typename from the TVP's full name */
+			pch = strchr(tvp_name, '.');
+			if (pch == NULL) {
+				schema = "";
+				typename = tvp_name;
+			} else {
+				*pch = 0;
+				schema = tvp_name;
+				typename = ++pch;
+			}
+
+			/* Ensure that the typename does not contain '.' */
+			pch = strchr(typename, '.');
+			if (pch != NULL) {
+				/* TODO: report error */
+			}
+
+			tvp = odbc_alloc_table(schema, typename);
+			if ((drec->sql_desc_data_ptr = (char *) tvp) == NULL) {
+				odbc_errs_add(&stmt->errs, "HY001", NULL);
+				ODBC_EXIT_(stmt);
+			}
+		}
+
+		/* field IPD related fields */
+		ipd = stmt->ipd;
+		orig_ipd_size = ipd->header.sql_desc_count;
+		if (ipar > ipd->header.sql_desc_count && desc_alloc_records(ipd, ipar) != SQL_SUCCESS) {
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY001", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		drec = &ipd->records[ipar - 1];
+
+		drec->sql_desc_parameter_type = fParamType;
+		if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS) {
+			desc_alloc_records(ipd, orig_ipd_size);
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY004", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		if (is_numeric) {
+			drec->sql_desc_precision = cbColDef;
+			drec->sql_desc_scale = ibScale;
+		} else {
+			drec->sql_desc_length = cbColDef;
+		}
 	} else {
-		drec->sql_desc_length = cbColDef;
+		tvp = (SQLTVP *) stmt->apd[stmt->param_focus - 1].records->sql_desc_data_ptr;
+		if (add_table_column(tvp, ipar, fParamType, fCType, fSqlType, cbColDef, ibScale, rgbValue, cbValueMax, pcbValue) != SQL_SUCCESS) {
+			odbc_errs_add(&stmt->errs, "HY001", NULL);
+			ODBC_EXIT_(stmt);
+		}
```

Comment:
If the current `param_focus` is set to a table, store the incoming values in the `SQLTVP` buffer instead of the statement.